### PR TITLE
Search plugin: Enable stats tracking upon establishing a site connection.

### PR DESCRIPTION
**Pull request by fgiannar**
This PR adds the stats tracking pixel for tracking pageviews, clicks etc upon establishing a site level connection in the Search plugin.
This is achieved by utilizing the Stats package and enabling the Stats module when:
1. The Search plugin is activated assuming an existing site level connection, or
2. A site level connection is established after activating the Search plugin

#### Changes proposed in this Pull Request:

* Adds the `stats` package dependency in the Search plugin
* Enables the Stats module upon establishing a site level connection by utilizing the `jetpack_site_registered` filter 
* Enables the Stats module upon plugin activation if the site is already connected 
* Utilizes the `jetpack_get_available_standalone_modules` filter, [required for standalone plugins](https://github.com/Automattic/jetpack/blob/b313bd253afdf919adf7a71a866d0ab85b3a7bda/projects/packages/status/src/class-modules.php#L258) to allow them to activate and deactivate modules even when Jetpack is not present.

#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

p9dueE-5TD-p2

#### Does this pull request change what data or activity we track or use?

Not sure here.. We track the same data/activity but on the Search plugin instead TBD

#### Testing instructions:

**Standalone Search plugin**
- On a JN site, without the JP plugin active, activate and connect the Search plugin
- Confirm pageviews are tracked by visiting any page in incognito mode and verifying you see the following:
Screenshot 2022-10-07 at 14 32 39

**Standalone Search plugin & Jetpack plugin**
- On your **local env** ([JN wont work here](https://github.com/Automattic/jetpack/pull/26697#issuecomment-1274613177))  activate and connect the Jetpack plugin
- Activate the Search plugin and verify pageviews are still tracked
- Play around with the Stats settings as displayed below and make sure they are respected:
Screenshot 2022-10-06 at 09 56 14
 
- Disable the stats module by visiting `wp-admin/admin.php?page=jetpack_modules` and confirm pageviews are no longer tracked
 - Deactivate the Jetpack plugin and confirm that pageviews are now tracked again
 
**Standalone Search plugin & any standalone plugin other than Jetpack**
- On a JN site activate and connect the Backup plugin
- Activate the Search plugin and verify pageviews are tracked
- Deactivate the Search plugin and confirm pageviews are no longer tracked

**Standalone Search plugin & old Jetpack plugin**
- On a JN site activate and connect an older version of the Jetpack plugin
- Activate the Search plugin and verify pageviews are still tracked **once**
- Play around with the Stats settings as displayed below and make sure they are respected

### CODE REVIEW

File: projects/plugins/search/src/class-jetpack-search-plugin.php [hunk 1]
```diff
@@ -27,7 +28,10 @@ public static function bootstrap() {
 		add_action( 'plugins_loaded', array( self::class, 'configure_packages' ), 1 );
 		add_action( 'plugins_loaded', array( self::class, 'initialize_other_packages' ) );
 		add_action( 'activated_plugin', array( self::class, 'handle_plugin_activation' ) );
+		add_action( 'activated_plugin', array( self::class, 'handle_plugin_activation' ) );
```
Comment (darssen) [inline projects/plugins/search/src/class-jetpack-search-plugin.php]: This line seems duplicated

**Review by darssen — DISMISSED** [review]
The code looks fine aside from one duplicated line.

### Testing

#### Standalone Search plugin

On a JN site, without the JP plugin active, activate and connect the Search plugin
- Confirm pageviews are tracked by visiting any page in incognito mode and verifying you see the following :heavy_check_mark: 
image

#### Standalone Search plugin & Jetpack plugin

On a JN site activate and connect the Jetpack plugin
- Activate the Search plugin and verify pageviews are still tracked:heavy_check_mark: 
- Play around with the Stats settings as displayed below and make sure they are respected: :heavy_check_mark: 
- Disable the stats module by visiting `wp-admin/admin.php?page=jetpack_modules `and confirm pageviews are no longer tracked :heavy_check_mark: 
- Deactivate the Jetpack plugin and confirm that pageviews are now tracked again :x:
    - This is not happening and I also found one disturbing behavior. When activating Jetpack plugin back again, I am unable to enable back the stats module unless I deactivate the Search Plugin. I tested with the Search plugin in the actual version and this behavior is not happening.
#### Standalone Search plugin & any standalone plugin other than Jetpack

On a JN site activate and connect the Backup plugin
- Activate the Search plugin and verify pageviews are tracked :heavy_check_mark: 
- Deactivate the Search plugin and confirm pageviews are no longer tracked :heavy_check_mark:

File: projects/plugins/search/src/class-jetpack-search-plugin.php (hunk 1 above)
Comment (fgiannar) [reply projects/plugins/search/src/class-jetpack-search-plugin.php]: 🤦  Fixed, thanks!

**Comment by fgiannar** [comment]
@darssen Many thanks for your review!

> Deactivate the Jetpack plugin and confirm that pageviews are now tracked again 

Should be fixed via https://github.com/Automattic/jetpack/pull/26697/commits/2155e9a0fe4392937800fbc6ef6ea023fb4bbd23

> When activating Jetpack plugin back again, I am unable to enable back the stats module unless I deactivate the Search Plugin.

I'm afraid I didn't manage to reproduce this. Could you please elaborate on the specific steps you took there?

**Comment by fgiannar** [comment]
> When activating Jetpack plugin back again, I am unable to enable back the stats module unless I deactivate the Search Plugin.

@darssen I was able to finally reproduce this with the scenario of old JP plugin version and the Search plugin version from this branch.
Essentially this means we have the old `stats.php` module file in the Jetpack plugin but also load the Stats package code.
In the Stats package code, we used to define `STATS_VERSION` constant (this functionality was moved from `stats.php` for backwards compatibility).
In the old version of the `stats.php`  however there’s this check: `if defined (STATS_VERSION) return` in the top of the file.
So, with no further delay, what’s happening here is that when `stats.php` code is loaded it returns early because the `STATS_VERSION` constant is already defined, therefore the module always appears as deactivated.
Btw I realized that `STATS_VERSION` is only used by the package now, I didn’t realize this earlier as we were extracting functionalities from `stats.php` bit by bit.
To resolve the issue, I turned it to a class constant instead

**Comment by darssen** [comment]
@fgiannar Really nice catch!

> @darssen I was able to finally reproduce this with the scenario of old JP plugin version and the Search plugin version from this branch.
> Essentially this means we have the old stats.php module file in the Jetpack plugin but also load the Stats package code.
> In the Stats package code, we used to define STATS_VERSION constant (this functionality was moved from stats.php for backwards compatibility).
> In the old version of the stats.php however there’s this check: if defined (STATS_VERSION) return in the top of the file.
> So, with no further delay, what’s happening here is that when stats.php code is loaded it returns early because the STATS_VERSION constant is already defined, therefore the module always appears as deactivated.

**Review by darssen — COMMENTED** [review]
Tested again and checked the code which looks fine for newer commits.
Still unable to pass the last scenario here.
#### Standalone Search plugin & Jetpack plugin

On a JN site activate and connect the Jetpack plugin
- Activate the Search plugin and verify pageviews are still tracked:heavy_check_mark: 
- Play around with the Stats settings as displayed below and make sure they are respected: :heavy_check_mark: 
- Disable the stats module by visiting `wp-admin/admin.php?page=jetpack_modules `and confirm pageviews are no longer tracked :heavy_check_mark: 
- Deactivate the Jetpack plugin and confirm that pageviews are now tracked again :x:
    - I am using the branch version for both JP plugin and Search plugin. If I am testing with 11.4 for JP and PR branch for Search it works fine.

**Comment by fgiannar** [comment]
> Tested again and checked the code which looks fine for newer commits. Still unable to pass the last scenario here.
> 
> #### Standalone Search plugin & Jetpack plugin
> On a JN site activate and connect the Jetpack plugin
> 
> * Activate the Search plugin and verify pageviews are still tracked✔️
> * Play around with the Stats settings as displayed below and make sure they are respected: ✔️
> * Disable the stats module by visiting `wp-admin/admin.php?page=jetpack_modules `and confirm pageviews are no longer tracked ✔️
> * Deactivate the Jetpack plugin and confirm that pageviews are now tracked again ❌
>   
>   * I am using the branch version for both JP plugin and Search plugin. If I am testing with 11.4 for JP and PR branch for Search it works fine.

This happens because the following will not work on on JN sites with JP branch testing:
`add_action( 'deactivate_jetpack/jetpack.php', array( self::class, 'activate_stats_module_on_deactivate_jetpack' ) );`
because when we deactivate jetpack the corresponding action is `deactivate_jetpack-dev/jetpack.php`

I have updated the testing instructions to test this on your local dev env instead.

**Comment by oskosk** [comment]
Rebased in [e51d1fd](https://github.com/Automattic/jetpack/pull/26697/commits/e51d1fdf084755e34f1cb27bbca6133a135f0827). A release happened to day and versions were off. Not sure if the rebase itself will help with version yet

**Review by darssen — DISMISSED** [review]
Tested locally instead of JN the troublesome scenario as suggested and works fine :)

**Review by coder-karen — COMMENTED** [review]
Tested all scenarios (including testing custom user roles) and everything worked well except for the last scenario (old Jetpack plugin version - I used 11.2), and this line in particular:

> Activate the Search plugin and verify pageviews are still tracked **once**

The pageviews were tracked each time, not just once, in that case. Same with v 10.0 (using the Search plugin version from this PR, on JN). Have I misunderstood that?

**Review by jsnmoon — APPROVED** [review]
The changes look good, and the described test cases behave as expected. :shipit: 

[EDIT: Deleted an irrelevant blurb about how I thought something was amiss but realized that I'd merely misunderstood the testing instructions.]

**Comment by jeherve** [comment]
I'm coming to this PR because of #29960. 

The issue above is problematic because Jetpack Stats **cannot be disabled** when using the VideoPress feature.

Things are a bit different with the Search feature, since we do not force-enable the Stats feature with stats; you can still disable it if you want to.

That said, I do not quite understand the purpose of this PR:

- What are we tracking in the Search feature with the Stats package, that is different from regular Jetpack Stats?
- Is it okay to enable the Stats feature (that will track visits on all pages of the site, make external requests, ...) on a site that enables the Search plugin, when the site owner on that site explicitly disabled the stats feature in the past? The module is automatically enabled in Jetpack when you connect the site to WordPress.com, so if it is deactivated there is a good reason for it, the site owner wanted that feature off.

**Comment by jeherve** [comment]
> What are we tracking in the Search feature with the Stats package, that is different from regular Jetpack Stats?

I found the answer to that question:

> A core part of the search algorithm is based off of stats from the site, so the search algorithm won't work very well if we don't have the stats module also in the plugin.
-- #24111

**Comment by cskindley** [comment]
> This PR adds the stats tracking pixel for tracking pageviews, clicks etc upon establishing a site level connection in the Search plugin. This is achieved by utilizing the Stats package and enabling the Stats module when:
> 
> 1. The Search plugin is activated assuming an existing site level connection, or
> 2. A site level connection is established after activating the Search plugin
> 
> #### Changes proposed in this Pull Request:
> * Adds the `stats` package dependency in the Search plugin
> * Enables the Stats module upon establishing a site level connection by utilizing the `jetpack_site_registered` filter
> * Enables the Stats module upon plugin activation if the site is already connected
> * Utilizes the `jetpack_get_available_standalone_modules` filter, [required for standalone plugins](https://github.com/Automattic/jetpack/blob/b313bd253afdf919adf7a71a866d0ab85b3a7bda/projects/packages/status/src/class-modules.php#L258) to allow them to activate and deactivate modules even when Jetpack is not present.
> 
> #### Other information:
> * [x]  Have you written new tests for your changes, if applicable?
> * [x]  Have you checked the E2E test CI results, and verified that your changes do not break them?
> 
> #### Jetpack product discussion
> p9dueE-5TD-p2
> 
> #### Does this pull request change what data or activity we track or use?
> Not sure here.. We track the same data/activity but on the Search plugin instead TBD
> 
> #### Testing instructions:
> **Standalone Search plugin**
> 
> * On a JN site, without the JP plugin active, activate and connect the Search plugin
> * Confirm pageviews are tracked by visiting any page in incognito mode and verifying you see the following:
> 
> Screenshot 2022-10-07 at 14 32 39
> 
> **Standalone Search plugin & Jetpack plugin**
> 
> * On your **local env** ([JN wont work here](https://github.com/Automattic/jetpack/pull/26697#issuecomment-1274613177))  activate and connect the Jetpack plugin
> * Activate the Search plugin and verify pageviews are still tracked
> * Play around with the Stats settings as displayed below and make sure they are respected:
> 
> Screenshot 2022-10-06 at 09 56 14
> 
> * Disable the stats module by visiting `wp-admin/admin.php?page=jetpack_modules` and confirm pageviews are no longer tracked
> * Deactivate the Jetpack plugin and confirm that pageviews are now tracked again
> 
> **Standalone Search plugin & any standalone plugin other than Jetpack**
> 
> * On a JN site activate and connect the Backup plugin
> * Activate the Search plugin and verify pageviews are tracked
> * Deactivate the Search plugin and confirm pageviews are no longer tracked
> 
> **Standalone Search plugin & old Jetpack plugin**
> 
> * On a JN site activate and connect an older version of the Jetpack plugin
> * Activate the Search plugin and verify pageviews are still tracked **once**
> * Play around with the Stats settings as displayed below and make sure they are respected

First I am not a developer. I am simply an end user of both free and paid JetPack Products. I do not use JetPack’s stats module. Never liked it and probably never will so it gets disabled as soon as possible upon installation on my business sites and client websites. If this idea that any product is enabled by default and overrides current settings because of a “dependency” —- strictly because YOU want it enabled to collect stats is asinine at best. It’s my website … not yours.

I shouldn’t have to rely on support providing a snippet workaround just to disable a “feature” that has long been able to be disabled without resorting to snippets.

I don’t like being critical but if I understand this intent it WILL cost a loss of paying customers. I use a different stats/analytics solution that makes forcing this to be on, with no way to disable it, resulted in higher resource usage, render blocking, drastic LCP times and poorer experience to my customers.

If I did misunderstand this please forgive me but this is my $0.02 worth. I simply respond with my wallet when good products screw things up.

**Pull request merged**